### PR TITLE
Optimize pfmatch

### LINKED ISSUE #240: Suboptimal pfmatch compilation
In pfmatch.md, there is the following example:

```lua
match {
   not ip => forward
   ip src 1.2.3.4 => incoming_ip(&ip[0])
   ip dst 5.6.7.8 => outgoing_ip(&ip[0])
   otherwise => drop
}
```

compiling to

```lua
local cast = require("ffi").cast
return function(self,P,length)
   if length < 14 then goto L5 end
   do
      if cast("uint16_t*", P+12)[0] ~= 8 then goto L5 end
      if length < 14 then goto L9 end
      do
         if cast("uint16_t*", P+12)[0] ~= 8 then goto L9 end
         if length < 34 then goto L9 end
         if cast("uint32_t*", P+26)[0] ~= 67305985 then goto L9 end
         return self.incoming_ip(P, len, 14)
      end
::L9::
      if length < 14 then goto L17 end
      do
         if cast("uint16_t*", P+12)[0] ~= 8 then goto L17 end
         if length < 34 then goto L17 end
         if cast("uint32_t*", P+30)[0] ~= 134678021 then goto L17 end
         return self.outgoing_ip(P, len, 14)
      end
::L17::
      return self.drop(P, len)
   end
::L5::
   return self.forward(P, len)
end
```

The "length<14" checks are repeated, needlessly.  The <34 checks should be hoisted.  The two duplicate checks at offset 12 should also be removed.

### PR DESCRIPTION
**Pull request by andywingo**
Fixes #240.

### CODE REVIEW

File: src/pf/optimize.lua [hunk 1]
```diff
@@ -724,19 +755,24 @@ local function lhoist(expr, db)
...
-      local min_expr = min_t
-      -- In tail position, only min_t counts.  Otherwise, restrict also
-      -- by min_f.  min_f can be nonzero only if a "fail" side effect
-      -- causes a subexpression to bail out early.
-      if not is_tail then min_expr = math.min(min_expr, min_f) end
-      if min < min_expr then
-         min = min_expr
-         local expr = reduce(aexpr, min, is_tail)
-         return { 'if', { '>=', 'len', min }, expr, { 'fail' } }
+      local min_t, min_f, min_pass, max_fail, expr =
```
Comment (mpeterv) [inline src/pf/optimize.lua]: min_f is unused?

**Comment by andywingo** [comment]
The code that's being output still isn't hitting the optimizations that we need.  Garr!

File: src/pf/optimize.lua (hunk 1 above)
Comment (andywingo) [inline src/pf/optimize.lua]: It's used when annotating.

**Comment by andywingo** [comment]
This is the IR coming out of the match expression from pfmatch.md:

```lua
{ "if",
  { "if",
    { ">=",
      "len",
      14 },
    { "=",
      { "[]",
        12,
        2 },
      8 },
    { "false" } },
  { "if",
    { "if",
      { ">=",
        "len",
        34 },
      { "if",
        { "=",
          { "[]",
            12,
            2 },
          8 },
        { "=",
          { "[]",
            26,
            4 },
          67305985 },
        { "false" } },
      { "false" } },
    { "call",
      "incoming_ip",
      14 },
    { "if",
      { "if",
        { ">=",
          "len",
          34 },
        { "if",
          { "=",
            { "[]",
              12,
              2 },
            8 },
          { "=",
            { "[]",
              30,
              4 },
            134678021 },
          { "false" } },
        { "false" } },
      { "call",
        "outgoing_ip",
        14 },
      { "call",
        "drop" } } },
  { "call",
    "forward" } }
```

**Comment by andywingo** [comment]
I think the challenge is that we are restricting ourselves to shrinking optimizations -- we can duplicate "fail" expressions but not "call" expressions.  That makes it hard for the algebraic optimizer to undo the nested "ifs".  Not sure, but I would consider duplicating these calls -- it might even be the same number of traces, considering that branches cause trace duplication.  Dunno.

**Comment by andywingo** [comment]
Scratch, that what we need is this:
```lua
(if (if T A false) TAIL1 (if (if T B false) TAIL3 TAIL4))
-> (if T (if A TAIL1 (if B TAIL3)) TAIL4)
```

No code growth.  I'll see what I can do.

**Comment by andywingo** [comment]
The real transformation:
```
if (if T A false) B (if (if T C false) D E)
-> if T (if A B (if C D E)) E
```

As you can see E is duplicated.  Hummm.

**Comment by andywingo** [comment]
If I allow calls to duplicate, then I get this:
```lua
local cast = require("ffi").cast
return function(self,P,length)
   if length >= 14 then
      if cast("uint16_t*", P+12)[0] == 8 then
         if length >= 34 then
            if cast("uint32_t*", P+26)[0] == 67305985 then
               return self.incoming_ip(P, len, 14)
            else
               if cast("uint32_t*", P+30)[0] == 134678021 then
                  return self.outgoing_ip(P, len, 14)
               else
                  return self.drop(P, len)
               end
            end
         else
            return self.drop(P, len)
         end
      else
         return self.forward(P, len)
      end
   else
      return self.forward(P, len)
   end
end
```

There is a way to avoid the duplication though if B and D are tail expressions, will see what happens there.

**Comment by kbara** [comment]
Duplication of E is avoidable, starting from if T (if A B (if C D E)) E:
```
(if (T&A) B 
   (if (!T&!A&C) D E))
```

**Comment by andywingo** [comment]
The way to avoid duplication goes like this:

```
if (if T A false) B (if (if T C false) D E)
-> if (if T (if A B (if C D false)) false) fail E
```

This works if B and D are tail expressions.  The fail is unreachable.  The results aren't great though:

```lua
local cast = require("ffi").cast
return function(self,P,length)
   if length < 14 then goto L5 end
   do
      if cast("uint16_t*", P+12)[0] ~= 8 then goto L5 end
      if length < 34 then goto L9 end
      do
         if cast("uint16_t*", P+12)[0] ~= 8 then goto L9 end
         if cast("uint32_t*", P+26)[0] == 67305985 then
            return self.incoming_ip(P, len, 14)
         end
         if cast("uint32_t*", P+30)[0] ~= 134678021 then goto L9 end
         return self.outgoing_ip(P, len, 14)
      end
::L9::
      return self.drop(P, len)
   end
::L5::
   return self.forward(P, len)
end
```

I expect we're bumping against the "scope tree is only a conservative approximation of a dominator tree" thing, so our CSE which runs over ANF isn't picking up things that it should.  The AST looks like this:

```lua
{ "if",
  { "if",
    { ">=",
      "len",
      14 },
    { "=",
      { "[]",
        12,
        2 },
      8 },
    { "false" } },
  { "if",
    { "if",
      { ">=",
        "len",
        34 },
      { "if",
        { "=",
          { "[]",
            12,
            2 },
          8 },
        { "if",
          { "=",
            { "[]",
              26,
              4 },
            67305985 },
          { "call",
            "incoming_ip",
            14 },
          { "if",
            { "=",
              { "[]",
                30,
                4 },
              134678021 },
            { "call",
              "outgoing_ip",
              14 },
            { "false" } } },
        { "false" } },
      { "false" } },
    { "fail" },
    { "call",
      "drop" } },
  { "call",
    "forward" } }
```

**Comment by andywingo** [comment]
kbara: as in many compiler ILs, we don't have "and" in this language, so unfortunately that transformation isn't valid.  Also the point here is to get T dominating A, B, C, and D, to enable other optimizations and that transformation doesn't allow that.

**Pull request merged**